corehq/apps/enterprise/models.py

```diff
@@ -63,3 +71,19 @@ class EnterpriseMobileWorkerSettings(models.Model):
     enable_auto_deactivation = models.BooleanField(default=False)
     inactivity_period = models.IntegerField(default=90)
     allow_custom_deactivation = models.BooleanField(default=False)
+
+    def deactivate_mobile_workers_by_inactivity(self, domain):
+        user_query = UserES().domain(domain).mobile_users().is_active()
+        user_data = user_query.source([
+            '_id',
+            'created_on',
+        ]).run()
+        for user in user_data.hits:
+            inactive_days = get_days_since_last_form_submission_for_user_id(
+                domain, user['_id']
+            )
+            if inactive_days == -1:
+                delta = datetime.datetime.utcnow() - iso_string_to_datetime(user['created_on'])
+                inactive_days = delta.days
+            if inactive_days >= self.inactivity_period:
+                auto_deactivate_commcare_user(user['_id'], domain)
```

Comment:
Similar to above, it seems inefficient to deactivate a single user at a time rather than collecting a batch of users to be deactivated, and then deactivating en-mass. Would that be possible?

Reply:
do you know of any instances where we currently make changes to users in bulk on HQ? even bulk upload seems to make changes one user at a time.

Reply:
ah good point. looks like `is_active` would fall under SQL. I will look into that tomorrow because I imagine the use case for this is for large-scale deployments.

Reply:
So I made an attempt at doing this. Unfortunately, while I can update the django `User` objects in bulk, it seems that the `CommCareUser` then becomes out of sync with the `User` object. I think the former way is much safer (though slower, unfortunately), especially since we need to create `UserHistory` objects for each of the deactivated users.

Here is my attempt. If you have any suggestions, please let me know:
```
def auto_deactivate_commcare_users(user_ids, domain):
    from corehq.apps.users.models import UserHistory, CommCareUser
    from corehq.apps.users.model_log import UserModelAction
    username_to_id = {}
    for user_id in user_ids:
        doc = CommCareUser.get_db().get(user_id)
        username_to_id[doc["username"]] = user_id
    user_query = User.objects.filter(
        username__in=username_to_id.keys(),
        is_active=True
    )
    active_usernames = user_query.values_list('username', flat='true')
    username_to_id_active = dict(
        (username, username_to_id[username])
        for username in active_usernames
    )

    # bulk deactivate users
    user_query.update(is_active=False)

    for username, user_id in username_to_id_active.items():
        UserHistory.objects.create(
            by_domain=domain,
            for_domain=domain,
            user_type=CommCareUser.doc_type,
            user_repr=username.split('@')[0],
            changed_by_repr=SYSTEM_USER_ID,
            user_id=user_id,
            changed_by=SYSTEM_USER_ID,
            changes={'is_active': False},
            changed_via=USER_CHANGE_VIA_AUTO_DEACTIVATE,
            change_messages={},
            action=UserModelAction.UPDATE.value,
            user_upload_record_id=None
        )
```

Reply:
> ```py
>     for user_id in user_ids:
>         doc = CommCareUser.get_db().get(user_id)
>         username_to_id[doc["username"]] = user_id
> ```
Is `user_id` stored in the SQL user model? If yes, it would probably be far more efficient to query these from SQL. I could be done with a single query rather than fetching each user object individually. If `user_id` is not in SQL maybe it would make sense to get `usernames` instead of `user_ids`, wherever those are coming from?

Reply:
I believe `user_id` is not stored in the `SQL` model as it uses the built-in `django.auth` `User` model, which is not storing couch ids.